core/sizes.cpp

```diff
@@ -2,81 +2,185 @@
+}
+
+template <class Key, class Value, int16_t Capacity>
+bool small_stable_map<Key, Value, Capacity>::operator!=(
+    const small_stable_map &other) const noexcept {
+  return !operator==(other);
+}
+
+template <class Key, class Value, int16_t Capacity>
+typename std::array<Key, Capacity>::const_iterator
+small_stable_map<Key, Value, Capacity>::find(const Key &key) const {
+  return std::find(begin(), end(), key);
+}
 
-scipp::index Sizes::at(const Dim dim) const {
-  scipp::expect::contains(*this, dim);
-  return m_sizes.at(dim);
+template <class Key, class Value, int16_t Capacity>
+bool small_stable_map<Key, Value, Capacity>::contains(
+    const Key &key) const noexcept {
+  return std::find(begin(), end(), key) != end();
 }
 
-scipp::index &Sizes::at(const Dim dim) {
-  scipp::expect::contains(*this, dim);
-  return m_sizes.at(dim);
+template <class Key, class Value, int16_t Capacity>
+scipp::index
+small_stable_map<Key, Value, Capacity>::index(const Key &key) const {
+  auto it = std::find(begin(), end(), key);
+  if (it == end())
+    throw_dimension_not_found_error(*this, key);
+  return std::distance(begin(), it);
 }
 
+template <class Key, class Value, int16_t Capacity>
+const Value &
+small_stable_map<Key, Value, Capacity>::operator[](const Key &key) const {
+  return at(key);
+}
+
+template <class Key, class Value, int16_t Capacity>
+const Value &small_stable_map<Key, Value, Capacity>::at(const Key &key) const {
+  if (!contains(key))
+    throw_dimension_not_found_error(*this, key);
+  return m_values[std::distance(begin(), find(key))];
+}
+
+template <class Key, class Value, int16_t Capacity>
+void small_stable_map<Key, Value, Capacity>::assign(const Key &key,
+                                                    const Value &value) {
+  if (!contains(key))
+    throw_dimension_not_found_error(*this, key);
+  m_values[std::distance(begin(), find(key))] = value;
+}
+
+template <class Key, class Value, int16_t Capacity>
+void small_stable_map<Key, Value, Capacity>::insert_left(const Key &key,
+                                                         const Value &value) {
+  expectUnique(*this, key);
+  expectExtendable(*this);
+  for (scipp::index i = size(); i > 0; --i) {
+    m_keys[i] = m_keys[i - 1];
+    m_values[i] = m_values[i - 1];
+  }
+  m_keys[0] = key;
+  m_values[0] = value;
+  m_size++;
+}
+
+template <class Key, class Value, int16_t Capacity>
+void small_stable_map<Key, Value, Capacity>::insert_right(const Key &key,
+                                                          const Value &value) {
+  expectUnique(*this, key);
+  expectExtendable(*this);
+  m_keys[m_size] = key;
+  m_values[m_size] = value;
+  m_size++;
+}
+
+template <class Key, class Value, int16_t Capacity>
+void small_stable_map<Key, Value, Capacity>::erase(const Key &key) {
+  if (!contains(key))
+    throw_dimension_not_found_error(*this, key);
+  m_size--;
+  for (scipp::index i = std::distance(begin(), find(key)); i < size(); ++i) {
```

Comment:
```suggestion
  for (scipp::index i = index(key); i < size()-1; ++i) {
```
Requires decrementing `m_size` after the loop. But allows dropping `if (!contains(key))`

Reply:
Nice!